string.c

```diff
@@ -3602,24 +3602,30 @@ str_uplus(VALUE str)
 
 /*
  * call-seq:
- *   -string -> frozen_string
- *   dedup -> frozen_string
+ *   -self -> frozen_string
  *
- * Returns a frozen, possibly pre-existing copy of the string.
+ * Returns a frozen string equal to +self+.
  *
- * The returned +String+ will be deduplicated as long as it does not have
- * any instance variables set on it and is not a String subclass.
+ * The return string is +self+ if and only if all of the following are true:
  *
- * Note that <tt>-string</tt> variant is more convenient for defining
- * constants:
+ * - +self+ is already frozen.
+ * - +self+ is an instance of \String (rather than of a subclass of \String)
+ * - +self+ has no instance variables set on it.
  *
- *    FILENAME = -'config/database.yml'
+ * Otherwise, the returned string is a frozen copy of +self+.
  *
- * while +dedup+ is better suitable for using the method in chains
- * of calculations:
+ * Returning +self+, when possible, saves duplicating data;
+ * see {Data deduplication}[https://en.wikipedia.org/wiki/Data_deduplication].
```

Comment:
While this is true, it's more powerful in that the returned string is such that for any other string `str` where `str == self`, then `-str` and `-self` will always be the same object. This allows for more deduplication that what this documentation implies. For example:

```ruby
str1 = "hello"
str2 = "hello"

puts "str1 and str2 are not the same object? #{str1.object_id != str2.object_id}"
puts "-str1 and -str2 are the same object? #{(-str1).object_id == (-str2).object_id}"
```

Reply:
Added.